PyInstaller/hooks/pre_find_module_path/hook-pyi_splash.py

```diff
@@ -0,0 +1,40 @@
+# -----------------------------------------------------------------------------
+# Copyright (c) 2005-2020, PyInstaller Development Team.
+#
+# Distributed under the terms of the GNU General Public License (version 2
+# or later) with exception for distributing the bootloader.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#
+# SPDX-License-Identifier: (GPL-2.0-or-later WITH Bootloader-exception)
+# -----------------------------------------------------------------------------
+
+"""
+This hook does not move a module that can be installed by a package manager,
+but points to a PyInstaller internal module that can be imported into
+the users python instance.
+
+The module is implemented in 'PyInstaller/fake-modules/pyi_splash.py'.
+"""
+
+import os
+
+from PyInstaller import PACKAGEPATH
+from PyInstaller.utils.hooks import logger
+
+
+def pre_find_module_path(api):
+    try:
+        # Test if a module named 'pyi_splash' is locally installed.
+        # This prevents that a potentially required dependency is not
+        # packed
+        import pyi_splash  # noqa: F401
+    except ImportError:
+        module_dir = os.path.join(PACKAGEPATH, 'fake-modules')
+
+        api.search_dirs = [module_dir]
+        logger.info('Adding pyi_splash module to application dependencies.')
+    else:
+        logger.info('A local module named "pyi_splash" is installed. '
+                    'Use the installed one instead.')
+        return
```

Comment:
My idea behind this check was, that I wanted to prevent issues in the future, where maybe someone installed a package named like `pyi_splash` (not necessarily a module to control the splash screen, but maybe a name collision with another project), which would be overwritten by our module. If such a module is installed I assumed the user explicitly installed and wanted to use it.

Reply:
Maybe we could change this to something like `PyInstaller.pyi_splash`? Update the hooks and rewrite the module a bit so that it just does nothing when unfrozen - so you can call it in unfrozen code. @bjones1 do you have an opinion?

Reply:
So, the core question is: where, conceptually, does the `pyi_splash` module belong? From what I understand, it is a module that will only be available in a frozen application and enables the frozen application to control the splash screen. The possibilities:

1. It is a fake module -- it replaces an existing system modules, such as `sites`. But this isn't correct, since it's not (or shouldn't be) overriding a system module.
2. It is a module that should be included as part of the modulegraph search process. However, it can't operate outside the frozen executable, so this isn't it.
3. It's a part of the loader, used to get the frozen Python environment bootstrapped. While another part of this PR does reside in the bootloader, this module isn't required for bootstrapping.
4. It's a part of the PyInstaller build process. No, because it runs when the frozen executable runs.
5. It's a run-time hook. These help set up the frozen environment for specific modules, and belong only in that context. While it's not required for splash screen operation (or is it -- will the splash screen close on its own?), this seems the closest description of what it does.

So, I think the name should be `pyi_rth_splash.py` or something similar, and it listed as a run-time hook.

Reply:
I thought about putting the module into the runtime-hook too, but I had some problems implementing it as one:

1. the runtime hook would need to be executed before the users script and after all other boot script, so it would need to be inserted into `Analysis.scripts` at a specific position after Analysis has finished. I think that would break very easily.

2. To still provide the module with its current name the script would need to modify `sys.modules` and do some weird loading. I think that would break very easily too.

I somewhat explained my choice here in [the previous PR.](https://github.com/pyinstaller/pyinstaller/pull/4581#discussion_r363060263)
Maybe you have a better way to implement this on the python side, but I didn't find any maintainable/satisfying way.

Reply:
I'm happy to leave it as a non-runtime hook if that fits more nicely. Is`pyi_splash` a module that the user imports to control the splash screen, which can only be imported/used in a frozen executable, and should therefore be included in the frozen executable?

Reply:
The `pyi_splash` module is to control the splash screen from within a frozen application. If someone tries to import it from a non frozen python interpreter the module cannot be found. So it only serves a purpose if the application is in a frozen state. If if it is not included in the frozen application and the bootloader starts a splash screen, there will be no way to close it.